bob-access/src/authenticator/basic.rs

```diff
@@ -1,65 +1,136 @@
-use std::{collections::HashMap, net::IpAddr};
+use std::{
+    collections::HashMap,
+    net::SocketAddr,
+    sync::{RwLock, Arc},
+    time::Duration,
+};
 
-use crate::{credentials::{Credentials, CredentialsKind}, AuthenticationType, error::Error, permissions::Permissions};
+use crate::{credentials::{DeclaredCredentials, RequestCredentials, CredentialsKind}, AuthenticationType, error::Error, permissions::Permissions};
 
 use super::{users_storage::UsersStorage, Authenticator};
 
 use sha2::{Digest, Sha512};
 
+use tokio::net::lookup_host;
+
+type NodesCredentials = HashMap<String, DeclaredCredentials>;
+
 #[derive(Debug, Default, Clone)]
 pub struct Basic<Storage: UsersStorage> {
     users_storage: Storage,
-    nodes: HashMap<IpAddr, Vec<Credentials>>,
+    nodes: Arc<RwLock<NodesCredentials>>,
+    resolve_sleep_dur_ms: u64,
 }
 
 impl<Storage: UsersStorage> Basic<Storage> {
-    pub fn new(users_storage: Storage) -> Self {
+    pub fn new(users_storage: Storage, resolve_sleep_dur_ms: u64) -> Self {
         Self {
             users_storage,
-            nodes: HashMap::new(),
+            nodes: Arc::new(RwLock::new(HashMap::new())),
+            resolve_sleep_dur_ms,
         }
     }
 
+    fn node_creds_ok(creds: &HashMap<String, DeclaredCredentials>) -> bool {
+        creds.values()
+            .all(|cred|
+                (cred.ip().is_some() || cred.hostname().is_some()) &&
+                cred.kind().map(|k| k.is_internode()) == Some(true))
+    }
+
     pub fn set_nodes_credentials(
         &mut self,
-        nodes: HashMap<IpAddr, Vec<Credentials>>,
+        mut nodes: HashMap<String, DeclaredCredentials>,
     ) -> Result<(), Error> {
-        if nodes
-            .values()
-            .all(|creds|
-                creds
-                    .iter()
-                    .all(|cred|
-                        cred.ip().is_some() &&
-                        cred.kind().map(|k| k.is_internode()) == Some(true)))
-        {
-            self.nodes = nodes;
+        if Self::node_creds_ok(&nodes) {
+            let mut resolved = HashMap::new();
+            for (nodename, cred) in nodes.drain() {
+                if cred.ip().is_none() && cred.hostname().is_some() {
+                    self.spawn_resolver(cred);
+                } else {
+                    resolved.insert(nodename, cred);
```

Comment:
Withe the current `resolve_worker` implementation, all nodes should be in `nodes_creds`